File: src-tauri/src/commands/decision.rs
use tauri::State;
use crate::state::AppState;
use crate::error::{AppError, AppResult};
use crate::services::decision_engine::{self, MasteryInfo};

/// 获取下一步教学动作
#[tauri::command]
pub async fn get_next_action(
    student_id: String,
    session_id: String,
    state: State<'_, AppState>,
) -> AppResult<serde_json::Value> {
    tracing::debug!("决策引擎: 学生 {} 会话 {}", student_id, session_id);

    // 从 DashMap 读取实时状态
    let (fatigue, consecutive_errors, session_secs) = if let Some(runtime) = state.student_states.get(&student_id) {
        (
            runtime.fatigue_level,
            runtime.consecutive_errors,
            runtime.session_duration_secs,
        )
    } else {
        (0.0, 0, 0)
    };

    let session_minutes = session_secs / 60;

    // 从数据库读取掌握度数据
    let mastery_data = {
        let db = state.db.lock().map_err(|e| AppError::Internal(e.to_string()))?;
        let mut mastery: Vec<MasteryInfo> = Vec::new();

        if let Ok(mut stmt) = db.prepare(
            "SELECT km.knowledge_id, COALESCE(kn.name, km.knowledge_id), km.mastery_score, km.forgetting_risk, km.last_practiced_at
             FROM knowledge_mastery km
             LEFT JOIN knowledge_nodes kn ON km.knowledge_id = kn.id
             WHERE km.student_id = ?1"
        ) {
            if let Ok(rows) = stmt.query_map(rusqlite::params![student_id], |row| {
                let last_practiced: Option<String> = row.get(4)?;
                let hours_ago = if let Some(ref lp) = last_practiced {
                    chrono::DateTime::parse_from_rfc3339(lp)
                        .map(|dt| {
                            let diff = chrono::Utc::now() - dt.with_timezone(&chrono::Utc);
                            diff.num_hours() as f64
                        })
                        .unwrap_or(0.0)
                } else {
                    0.0
                };
                Ok(MasteryInfo {
                    knowledge_id: row.get(0)?,
                    name: row.get(1)?,
                    mastery_score: row.get(2)?,
                    forgetting_risk: row.get(3)?,
                    last_practiced_hours_ago: hours_ago,
                })
            }) {
                for row in rows.flatten() {
                    mastery.push(row);
                }
            }
        }
        mastery
    };

    // 调用决策引擎
    let result = decision_engine::decide(
        fatigue,
        0.0, // frustration 暂未独立计算
        consecutive_errors,
        session_minutes,
        30, // max_session_minutes
        &mastery_data,
    );

    Ok(serde_json::json!({
        "action": result.action,
        "reasoning": result.reasoning,
        "params": result.params,
        "context": {
            "fatigue": fatigue,
            "consecutive_errors": consecutive_errors,
            "session_minutes": session_minutes,
            "mastery_count": mastery_data.len(),
        }
    }))
}

File: server/app/routes/decision.py
"""决策路由 - 调 decision_engine"""
import aiosqlite
from fastapi import APIRouter, Depends

from ..db.connection import get_db
from ..schemas.common import ok
from ..schemas.decision import NextActionRequest
from ..services.decision_engine import MasteryInfo, decide
from ..services.student_model import forgetting_risk
from ._deps import require_api_key

router = APIRouter(dependencies=[Depends(require_api_key)])


@router.post("/decision/next")
async def next_action(
    payload: NextActionRequest, db: aiosqlite.Connection = Depends(get_db)
):
    # 读学生当前状态
    async with db.execute(
        "SELECT * FROM student_states WHERE student_id = ?",
        (payload.student_id,),
    ) as cur:
        state_row = await cur.fetchone()

    fatigue = float(state_row["fatigue"]) if state_row else 0.0
    frustration = float(state_row["frustration"]) if state_row else 0.0
    consecutive_errors = int(state_row["consecutive_errors"]) if state_row else 0

    # 会话时长
    session_minutes = 0
    if payload.session_id:
        async with db.execute(
            """
            SELECT
              CAST((julianday('now') - julianday(started_at)) * 24 * 60 AS INTEGER) AS mins
            FROM learning_sessions WHERE id = ?
            """,
            (payload.session_id,),
        ) as cur:
            row = await cur.fetchone()
            if row:
                session_minutes = int(row["mins"] or 0)

    # 读掌握度
    async with db.execute(
        """
        SELECT km.knowledge_id, k.name, km.mastery_score,
               CAST((julianday('now') - julianday(COALESCE(km.last_practiced_at, 'now'))) * 24 AS REAL) AS hours_ago
        FROM knowledge_mastery km
        JOIN knowledge_nodes k ON k.id = km.knowledge_id
        WHERE km.student_id = ?
        """,
        (payload.student_id,),
    ) as cur:
        rows = await cur.fetchall()

    mastery = [
        MasteryInfo(
            knowledge_id=r["knowledge_id"],
            name=r["name"],
            mastery_score=float(r["mastery_score"]),
            forgetting_risk=forgetting_risk(
                float(r["mastery_score"]), float(r["hours_ago"] or 0)
            ),
            last_practiced_hours_ago=float(r["hours_ago"] or 0),
        )
        for r in rows
    ]

    result = decide(
        fatigue=fatigue,
        frustration=frustration,
        consecutive_errors=consecutive_errors,
        session_minutes=session_minutes,
        max_session_minutes=30,
        mastery_data=mastery,
    )
    return ok(
        {
            "action": result.action,
            "reasoning": result.reasoning,
            "params": result.params,
        }
    )
